backend/queues/makeOfferProcessor.js

```diff
@@ -34,48 +57,126 @@ async function processor(job) {
+    // It's possible the processing got interrupted while waiting for the tx to get mined.
+    // For example due to a server maintenance or a crash.
+    transaction = await Transaction.findOne({
+      where: {
+        networkId: network.networkId,
+        wallet: walletAddress,
+        status: TransactionStatuses.Pending
+      }
+    })
+    if (transaction) {
+      log.info(
+        `Found pending transaction ${transaction.id} job ${job.id} hash ${transaction.hash} for wallet ${walletAddress}`
+      )
+
+      // If it is not the transaction from our job. Do not try to recover it.
+      // Let it get recovered by the job that created it.
+      // Fail our job for now, it will get retried.
+      if (transaction.jobId === job.id) {
+        throw new Error(
+          `Pending transaction does not belongs to job ${job.id}. Bailing.`
+        )
+      }
+
+      // Try to recover by loading the tx from its hash.
+      tx = await provider.getTransaction(transaction.hash)
+      if (!tx) {
+        // The transaction was not mined and is not in the transaction pool.
+        // Something went really wrong...
+        throw new Error(
+          `Transaction ${transaction.id} with hash ${transaction.hash} not found`
+        )
+      }
+      log.info('Recovered tx', tx)
+    } else {
+      // Send a blockchain transaction to make an offer on the marketplace contract.
+      queueLog(25, 'Sending to marketplace')
+      tx = await _createOffer(
+        provider,
+        marketplace,
+        wallet,
+        lid,
+        offer,
+        ipfsHash
+      )
+      log.info('Transaction sent:', tx)
+
+      // Record the transaction in the DB.
+      transaction = await Transaction.create({
+        shopId,
+        networkId: network.networkId,
+        wallet: walletAddress,
+        type: TransactionTypes.OfferCreated,
+        status: TransactionStatuses.Pending,
+        hash: tx.hash,
+        listingId: lid.toString(),
+        ipfsHash,
+        jobId: job.id.toString()
+      })
+    }
+
+    // Wait for the tx to get mined.
+    // Note: this is blocking with no timeout. Depending on the network conditions,
+    // it could take a long time for a tx to get mined. This will become a bottleneck
+    // in the future when the transaction volume scales up. A potential solution
+    // will be to run the confirmation logic as a separate queue with multiple workers.
+    queueLog(50, `Waiting for tx ${tx.hash} to get confirmed`)
+    log.info('Waiting for tx confirmation...')
+    confirmation = await _waitForMakeOfferTxConfirmation(marketplace, tx)
+    const { receipt, offerId } = confirmation
+
+    // Update the transaction in the DB.
+    const oid = new OfferID(lid.listingId, offerId, network.networkId)
+    await transaction.update({
+      status: TransactionStatuses.Confirmed,
+      blockNumber: receipt.blockNumber,
+      offerId: oid.toString() // Store the fully qualified offerId.
```

Comment:
Can we guarantee offer ID is unique in the `transaction` table?

Reply:
In the future we might have more transaction types recorded. For example OfferWithdrawal/Acceptance/Finalization and in that case offerId won't be unique. So I think I'll leave it as is for now.